guide/src/reference/reference-types.md

```diff
@@ -0,0 +1,59 @@
+# Support for Reference Types
+
+WebAssembly recently has gained support for a new value type called `externref`.
+Proposed in the [WebAssembly reference types
+repo](https://github.com/webassembly/reference-types) this feature of
+WebAssembly is hoped to enable more efficient communication between the host
+(JS) and the wasm module. This feature removes the need for much of the JS glue
+generated by `wasm-bindgen` because it can natively call APIs with JS values.
+
+For example, this Rust function:
+
+```rust
+#[wasm_bindgen]
+pub fn takes_js_value(a: &JsValue) {
+    // ...
+}
+```
+
+generates this JS glue *without* reference types support:
+
+```js
+const heap = new Array(32).fill(undefined);
+
+heap.push(undefined, null, true, false);
+
+let stack_pointer = 32;
+
+function addBorrowedObject(obj) {
+    if (stack_pointer == 1) throw new Error('out of js stack');
```

Comment:
Hey quick question.

How come this isn't `if (stack_pointer == 0)`?

What's the issue with assigning an object to `heap[0]`.

Just for my own understanding.

Thanks!

Reply:
I think at some point we were reserving the first slot (index 0) as a sort of null pointer or something like that. I don't believe those plans ever panned out though so this is likely a bug that we can fix nowadays.

Reply:
Sweet thanks for explaining!